python_onnxifi/python_onnxifi.cc

```diff
@@ -0,0 +1,355 @@
+  BackendRep(std::string&& serializedModel, onnxBackend backend) : graph_(), backend_(backend), serialized_model_(serializedModel), conversion_() {
+    if (onnxInitGraph(backend, serialized_model_.size(), serialized_model_.c_str(), 0, nullptr, &graph_)
+                                         != ONNXIFI_STATUS_SUCCESS) {
+      std::cerr << "Could not initialize graph on given device" << std::endl;
+      graph_ = NULL; //Make dummy BackendRep with NULL graph TODO: Return python None object
+    }
+  }
+
+  // TODO: Implement run
+  //TODO: Change usage of memoryFence (onnxEvent instead of onnxEvent*) when submodule pulled
+  py::dict run(py::dict inputs, py::kwargs kwargs) {
+    conversion_.updateInputDescriptorsData(inputs);
+    auto inputDescriptors = conversion_.getInputTensorDescriptors();
+    ModelProto model;
+    if (!model.ParseFromString(serialized_model_))  {
+      std::cerr << "Failed to parse model proto" << std::endl;
+      //TODO: Return None
+    }
+    conversion_.updateOutputDescriptorsData(model);
+    auto outputDescriptors = conversion_.getOutputTensorDescriptors();
+    if (onnxSetGraphIO(graph_, inputDescriptors.size(), inputDescriptors.data(),
+                               outputDescriptors.size(), outputDescriptors.data()) != ONNXIFI_STATUS_SUCCESS)  {
+      std::cerr << "I/O could not be set" << std::endl;
+      //TODO: Return None
+    }
+    onnxMemoryFence inputFence;
+    inputFence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+    onnxEvent inputEvent;
+    if (onnxInitEvent(backend_, &inputEvent) != ONNXIFI_STATUS_SUCCESS)  {
+      throw std::runtime_error("Internal Error: Event for input memory fence could not be initialized (expected ONNXIFI_STATUS_SUCCESS)");
+    }
+    if (onnxSignalEvent(inputEvent) != ONNXIFI_STATUS_SUCCESS)  {
+      throw std::runtime_error("Internal Error: Event for input memory fence could not be signalled (expected ONNXIFI_STATUS_SUCCESS)");
+    }
+    inputFence.event = &inputEvent;
+    onnxMemoryFence outputFence;
+    outputFence.type = ONNXIFI_SYNCHRONIZATION_EVENT;
+    onnxEvent outputEvent;
+    outputFence.event = &outputEvent;
+    if (onnxRunGraph(graph_, &inputFence, &outputFence) != ONNXIFI_STATUS_SUCCESS)  {
+      std::cerr << "Graph failed to run successfully" << std::endl;
+      //TODO: Return None
+    }
+    if (onnxWaitEvent(outputEvent) != ONNXIFI_STATUS_SUCCESS)  {
+      throw std::runtime_error("Internal Error: Error waiting for output event (expected ONNXIFI_STATUS_SUCCESS)");
+    }
+    if (onnxReleaseEvent(inputEvent) != ONNXIFI_STATUS_SUCCESS)  {
+      throw std::runtime_error("Internal Error: Error releasing input event (expected ONNXIFI_STATUS_SUCCESS)");
+    }
+    if (onnxReleaseEvent(outputEvent) != ONNXIFI_STATUS_SUCCESS)  {
+      throw std::runtime_error("Internal Error: Error releasing output event (expected ONNXIFI_STATUS_SUCCESS)");
+    }
+
+    return conversion_.getNumpyOutputs();    
```

Comment:
IIUC conversion_ is right now a combination of:

1. several conversion helper functions
2. a container for holding the input, output and weight descriptors. However, input and output descriptors are always get recomputed in each run so no need to live outside of this `run` function, and weight descriptors are never used.

How about change those helper functions to return the descriptors instead of storing them as member variables in conversion_ and get them immediately, i.e.
```c++
// do this 
auto inputDescriptors = conversion::GetInputTensorDescriptors(inputs);

// instead of
conversion_.updateInputDescriptorsData(inputs);
auto inputDescriptors = conversion_.getInputTensorDescriptors();
```

Reply:
I think the above sort of compression into one line of code is a good point - no need to call the updateInputDescriptorsData function externally. However, I think member variables containers for the DescriptorData objects are necessary. First, they are used for memory management - the shape, name, and buffer in the onnxTensorDescriptors all need some sort of management on the heap. The approach with member variables will keep the string/vector containers alive while the onnxTensorDescriptors are used. Second, we shouldn't actually be refilling the descriptors every run (I put some comments in the header file explaining how to only update descriptors when necessary, but I suppose we can change in this PR itself).

Here is my proposal of what to modify:

1. In terms of IO descriptors, we can have one public function

`bool updateIOTensorDescriptors(py::dict inputs, vector<onnxTensorDescriptor>& inputDescriptors, vector<onnxTensorDescriptor>& outputDescriptors)`

This function will return the input and output descriptors (by reference) to the BackendRep run function. The bool return value will indicate if in the update operation any of the tensorDescriptor metadata was changed - that is, did any of the I/O descriptor data change other than the data stored input buffers. If